generator/src/main/java/com/scottlogic/deg/generator/generation/DecisionTreeDataGenerator.java

```diff
@@ -45,37 +46,35 @@ public DecisionTreeDataGenerator(
         DecisionTree decisionTree,
         GenerationConfig generationConfig) {
 
-        final List<DecisionTree> partitionedTrees =
-            treePartitioner
-                .splitTreeIntoPartitions(decisionTree)
-                    .map(this.treeOptimiser::optimiseTree)
-                .collect(Collectors.toList());
+        monitor.generationStarting(generationConfig);
+        CombinationStrategy partitionCombiner = generationConfig.getCombinationStrategy();
 
-        final Stream<Stream<RowSpec>> rowSpecsByPartition = partitionedTrees
-            .stream()
-            .map(tree -> treeWalker.walk(tree,
-                walkerStrategyFactory.getWalkerStrategy(profile, tree, generationConfig)));
+        Stream<Stream<DataBag>> partitionedDataBags = treePartitioner
+            .splitTreeIntoPartitions(decisionTree)
+            .map(treeOptimiser::optimiseTree)
+            .map(tree -> generateDataBags(profile, tree, generationConfig));
 
-        final Stream<DataBagSource> allDataBagSources =
-            rowSpecsByPartition
-                .map(rowSpecs ->
-                    new ConcatenatingDataBagSource(
-                        rowSpecs
-                            .map(dataBagSourceFactory::createDataBagSource)));
+        return partitionCombiner.permute(partitionedDataBags)
+            .map(dataBag -> toGeneratedObject(profile, dataBag))
+            .limit(generationConfig.getMaxRows().orElse(GenerationConfig.Constants.DEFAULT_MAX_ROWS))
+            .peek(monitor::rowEmitted);
+    }
 
-        Stream<GeneratedObject> dataRows = new MultiplexingDataBagSource(allDataBagSources)
-            .generate(generationConfig)
-            .map(dataBag -> new GeneratedObject(
-                profile.fields.stream()
-                    .map(dataBag::getValueAndFormat)
-                    .collect(Collectors.toList()),
-                dataBag.getRowSource(profile.fields)));
+    private Stream<DataBag> generateDataBags(Profile profile, DecisionTree tree, GenerationConfig config) {
```

Comment:
I think the methods in this type can be better named. A method named `generateDataBags` describes the consequence more than the reason/intent.

If this method was renamed to something like `generateData` or `generateDataForTree` it would convey the meaning that it is responsible for generating data, not DataBags (albeit they're the vehicle to expose the data).

same goes for `toGeneratedObject`, a better name would really help people understand the point on the method.

Something like `convertToOuputType` (not the best, but conveys the reason for the mapping at least)

Reply:
`generateForPartition` & `converToGeneratedObject`